### ISSUE TITLE
anonymous functions in `map` still very slow

### ISSUE THREAD

**Issue by ExpandingMan** [JuliaLang/julia#18874]
Anonymous functions still seem to be incredibly slow (by factors of between 10 and 50) when used in `map` if they are not declared first.

```julia
A = rand(10^6)

@time α = map(x -> x^2, A)
>>> 0.040819 seconds (13.36 k allocations: 8.220 MB)

f(x) = x^2
@time β = map(f, A)
>>> 0.003943 seconds (8 allocations: 7.630 MB)

@time γ = A.^2
>>> 0.009254 seconds (6 allocations: 7.630 MB)
```
(These numbers are from the second time the code runs, so should not include compilation in any case.)
Note that `do` blocks seem similarly slow.  I'm also surprised that `A.^2` seems significantly slower than `map(f, A)`.

I got the impression from the v0.5 changelog that this shouldn't still be happening, so I thought I'd raise the issue.

**Comment by StefanKarpinski** [comment]
That's an artifact of the way you're timing – since you're creating a new function object each time, the new function gets JITed every time. In real code, you won't be typing in the same closure every time so this doesn't have any impact. This is why the Julia 0.5 highlights post either a) wraps the usage of map in another function or b) uses BenchmarkTools to measure performance:

```jl
julia> using BenchmarkTools

julia> @benchmark map(x->x^2, A)
BenchmarkTools.Trial:
  samples:          957
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  7.63 mb
  allocs estimate:  4
  minimum time:     2.93 ms (0.00% GC)
  median time:      3.93 ms (0.00% GC)
  mean time:        5.22 ms (27.52% GC)
  maximum time:     21.27 ms (60.72% GC)

julia> @benchmark map(f, A)
BenchmarkTools.Trial:
  samples:          952
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  7.63 mb
  allocs estimate:  4
  minimum time:     2.93 ms (0.00% GC)
  median time:      4.01 ms (0.00% GC)
  mean time:        5.24 ms (27.84% GC)
  maximum time:     16.21 ms (55.54% GC)

julia> @benchmark A.^2
BenchmarkTools.Trial:
  samples:          579
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  7.63 mb
  allocs estimate:  3
  minimum time:     6.44 ms (0.00% GC)
  median time:      7.35 ms (0.00% GC)
  mean time:        8.64 ms (15.50% GC)
  maximum time:     16.64 ms (28.65% GC)
```

**Comment by ExpandingMan** [comment]
Hmm... I'm still a little confused, the `macroexpand` of time shows it sandwiching `map` in a timer.  Since `map` runs normally I would have the the JIT would just proceed normally on the line that says `map`.  Anyway, thanks for the explanation.

**Comment by ExpandingMan** [comment]
Oh... I just realized what you meant.  It gets recompiled every time `@time` gets called.  That makes sense.  Thanks for the clarification.

**Comment by StefanKarpinski** [comment]
No problem. Ideally we'll be able to cache the JITed code for anonymous functions like this, but knowing when two of them are equivalent is a little hard and since it doesn't in practice matter because real code doesn't repeat the same anonymous function over and over again, it's not a big priority – the effect is only visible in the REPL and only matters and even there, for long-running code the compile time is a trivial proportion of the execution.

**Comment by vtjnash** [comment]
We will probably need to fix this for v0.6, since it causes problems for generated functions. I'm not certain if we have another issue open for it?
